src/transform/src/attribute/non_negative.rs

```diff
@@ -0,0 +1,221 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Definition and helper structs for the [`NonNegative`] attribute.
+
+use std::collections::HashMap;
+
+use mz_expr::visit::Visitor;
+use mz_expr::MirRelationExpr;
+use mz_expr::{Id, LocalId};
+
+/// Traverses a [`MirRelationExpr`] tree and figures out whether for subtree
+/// the sum of all diffs up to a specific time for any record can be a
+/// negative value.
+///
+/// The sresult for each subtree are accumulated in a bottom-up order
+/// in [`NonNegative::results`].
+///
+/// This method is a conservative approximation and is known to miss not-hard
+/// cases.
+///
+/// It assumes that all `Get` bindings correspond to collections without negative
+/// multiplicities. Local let bindings present in `safe_lets` are relied on to have
+/// no non-negative multiplicities.
+///
+/// The [`Visitor`] implementation handles the management of the results environment.
+#[derive(Default)]
+#[allow(missing_debug_implementations)]
+pub struct NonNegative {
+    /// A map that keeps the computed `result` value for all `LocalId`
+    /// bindings in the current environment.
+    ///
+    /// The relevant result is always the last entry in the associated
+    /// `Vec<bool>` vector, while entires keep the results of shadowed
+    /// values. In principle we don't allow shadowing, but this makes
+    /// the code more robust until we build higher confidence that
+    /// this indeed is the case also in practice.
+    env: HashMap<LocalId, bool>,
+    // A stack of tasks to maintain the `env` map.
+    env_tasks: Vec<EnvTask>,
+    /// A stack of results for all nodes in the visited tree in
```

Comment:
I see only pushes to it. If there are indeed no pops, then I wouldn't call it a stack.

Reply:
Agreed!